app/routes.py

```diff
@@ -1,2 +1,172 @@
+            tasks = Task.query.order_by(asc(Task.title))
+        elif sort_query == "desc":
+            tasks = Task.query.order_by(desc(Task.title))
+        else:
+            return make_response({"details": "Invalid query"}, 400)
+    else:
+        tasks = Task.query.all()
+    if request.method == "GET":
+        task_response = []
+
+        for task in tasks:
+            task_response.append(task.to_json())
+        return jsonify(task_response), 200
+    elif request.method == "POST":
+        request_body = request.get_json()
+        try:
+            new_task = Task.new_task_from_json(request_body)
+        except KeyError:
+            return make_response({"details": "Invalid data"}, 400)
+        db.session.add(new_task)
+        db.session.commit()
+
+        return make_response({"task": new_task.to_json()}, 201)
+
+@task_bp.route("/<task_id>", methods=["GET", "PUT", "DELETE"])
+def handle_single_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return return_404()
+    if request.method == "GET":
+        return {"task": task.to_json()}
+    elif request.method == "PUT":
+        request_body = request.get_json()
+        try:
+            new_task = Task.new_task_from_json(request_body)
+            # task.title = request_body["title"]
+            # task.description = request_body["description"]
+            # task.completed_at = None
+        except KeyError:
+            return make_response({"details": "Invalid data"}, 400)
+
+        db.session.commit()
+        updated_task = new_task.to_json()
+        updated_task["is_complete"] = task.is_complete()
+
+        return make_response({"task": updated_task}, 200)
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        return make_response({"details": f'Task {task.task_id} "{task.title}" successfully deleted'}, 200)
+
+
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def update_task_with_completion(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return return_404()
+    else: 
+        task.completed_at = date.today()
+        db.session.commit()
+        updated_task = task.to_json()
+        updated_task["is_complete"] = task.is_complete()
+        token = os.environ.get("SLACK_BOT_TOKEN")
+        message = {"channel": "task-notifications", "text": "Someone just completed the task " + str(task.title)}
+        slackbot_request = requests.post("https://slack.com/api/chat.postMessage", json=message,  headers={"Authorization": token})
+        
+    return{"task": updated_task}
+
+
+@task_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def update_task_with_incomplete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return return_404()
+    else: 
+        updated_task = task.to_json()
+        task.completed_at = None
+        db.session.commit()
+        updated_task["is_complete"] = task.is_complete()
```

Comment:
Putting the lines in this order makes them a little clearer 😄:  

```suggestion
        task.completed_at = None
        db.session.commit()
        
        updated_task = task.to_json()
        updated_task["is_complete"] = task.is_complete()
```